### PR TITLE
[Refactor] Introduce ColumnId to support Column renaming (part4)

### PR DESCRIPTION
**Pull request by gengjun-git**
## Why I'm doing:
We need a unique ID to identify the Column. This ID is used in all places that reference the Column. In this way, to change the attributes of the Column, such as name, we only need to change the attributes in the Column object. There is a uniqueId in the current Column, but it is only available in newly created tables. For compatibility reasons, we need to introduce another Id: ColumnID. The ColumnID of the historical table is the name of the column, because the name was previously immutable.

## What I'm doing:
There are currently three ways to reference Column: 1: direct copy of Column object, 2: reference to Column name, 3: sql expression reference.
This PR changes the sql expression reference to use ColumnId reference.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            slotRef.setColumnId(column.getColumnId());
+        }
+
+        for (Expr child : expr.getChildren()) {
+            setColumnId(nameToColumn, child);
+        }
+    }
+
+    private static void setColumnIdToColumnName(Expr expr) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java:96]:
```suggestion
    private static void setColumnIdByColumnName(Expr expr) {
```

better?

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        return this.expr.equals(((ColumnIdExpr) obj).expr);
+    }
+
+    private static class ExprSerializeVisitor extends AstToStringBuilder.AST2StringBuilderVisitor {
+        @Override
+        public String visitSlot(SlotRef node, Void context) {
+            if (node.getTblNameWithoutAnalyzed() != null) {
+                return node.getTblNameWithoutAnalyzed().toString() + "." + node.getColumnId().getId();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java:124]: why use column id here, should use column name?

File: fe/fe-core/src/main/java/com/starrocks/sql/common/MetaUtils.java.rej [hunk 3]
```diff
@@ -0,0 +1,120 @@
+diff a/fe/fe-core/src/main/java/com/starrocks/sql/common/MetaUtils.java b/fe/fe-core/src/main/java/com/starrocks/sql/common/MetaUtils.java	(rejected hunks)
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/MetaUtils.java.rej:1]: remove this file?

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        Expr clonedExpr = expr.clone();
+        setColumnName(MetaUtils.convertToIdToColumn(schema), clonedExpr);
+        return clonedExpr;
+    }
+
+    public Expr getExpr() {
+        return expr;
+    }
+
+    public String serialize() {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java:56]: why call it serialize/deserialize? how about toSQL/fromSQL?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Column.java [hunk 5]
```diff
@@ -252,7 +254,7 @@ public ColumnDef toColumnDef() {
             }
         }
         ColumnDef col = new ColumnDef(name, new TypeDef(type), null, isKey, aggregationType, isAllowNull,
-                defaultValueDef, isAutoIncrement, generatedColumnExpr, comment);
+                defaultValueDef, isAutoIncrement, generatedColumnExpr.convertToColumnNameExpr(table.getIdToColumn()), comment);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Column.java:257]: why using column name here?

File: fe/fe-core/src/main/java/com/starrocks/catalog/ExpressionRangePartitionInfo.java [hunk 6]
```diff
@@ -173,60 +145,15 @@ public String toSql(OlapTable table, List<Long> partitionId) {
             sb.append(")");
             return sb.toString();
         }
-        sb.append(Joiner.on(", ").join(partitionExprs.stream().map(Expr::toSql).collect(toList())));
+        sb.append(Joiner.on(", ").join(partitionExprs
+                .stream()
+                .map(physicalExpr -> physicalExpr.convertToColumnNameExpr(table.getIdToColumn()).toSql())
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ExpressionRangePartitionInfo.java:150]:
```suggestion
                .map(columnIdExpr -> columnIdExpr.convertToColumnNameExpr(table.getIdToColumn()).toSql())
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/SlotRef.java [hunk 7]
```diff
@@ -62,7 +63,8 @@
 
 public class SlotRef extends Expr {
     private TableName tblName;
-    private String col;
+    private String colName;
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/analysis/SlotRef.java:66]: yes

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java (hunk 2 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java:122]: this is used when ColumnIdExpr is serializing to store, we should use ColumnId.

File: fe/fe-core/src/main/java/com/starrocks/persist/ExpressionSerializedObject.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+public class ExpressionSerializedObject {
+    public ExpressionSerializedObject(String expressionSql) {
+        this.expressionSql = expressionSql;
+    }
+
+    @SerializedName("expr")
+    private String expressionSql;
+
+    public static ExpressionSerializedObject create(ColumnIdExpr expr) {
+        return new ExpressionSerializedObject(expr.toSql());
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/persist/ExpressionSerializedObject.java:29]: check expr

File: fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import com.starrocks.qe.SqlModeHelper;
+import com.starrocks.sql.analyzer.AstToStringBuilder;
+import com.starrocks.sql.analyzer.SemanticException;
+import com.starrocks.sql.common.MetaUtils;
+import com.starrocks.sql.parser.SqlParser;
+
+import java.util.List;
+import java.util.Map;
+
+public class ColumnIdExpr {
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/analysis/ColumnIdExpr.java]: done

File: fe/fe-core/src/main/java/com/starrocks/persist/ExpressionSerializedObject.java (hunk 8 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/persist/ExpressionSerializedObject.java]: done

**Review by nshangyiming — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Comment by gengjun-git** [comment]
@mergifyio backport branch-3.3

**Pull request merged**